feat(core): CfnResource depends on methods

jusdino opened the pull request:
Add some new methods to allow a minimal interface for viewing and editing resource-to-resource dependencies that mirrors the behavior of `CfnResource.addDependsOn()`.

Related to #20418 - details for justification are there

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)
* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)
* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

### Adding new Unconventional Dependencies:


### New Features


*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: ac8424700d112973304a8bbbd68254deb66905d9
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f0cac7f8-bb5a-485e-949d-d0e42711cf39%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

comcalvi commented on packages/@aws-cdk/core/test/cfn-resource.test.ts:
```diff
@@ -50,6 +50,78 @@ describe('cfn resource', () => {
+      resource1.removeDependsOn(resource2);
+      expect(stack1.dependencies).toEqual([]);
+    });
+
+
```
Please delete these two blank lines

comcalvi commented on packages/@aws-cdk/core/test/cfn-resource.test.ts:
```diff
@@ -50,6 +50,78 @@ describe('cfn resource', () => {
+          },
+        },
+      });
+
+
```
Please delete these two blank lines

comcalvi commented on packages/@aws-cdk/core/test/cfn-resource.test.ts:
```diff
@@ -50,6 +50,78 @@ describe('cfn resource', () => {
+          },
+        },
+      });
+
```
Please delete this blank line

comcalvi commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -261,6 +261,31 @@ export class CfnResource extends CfnRefElement {
+  /**
+   * Indicates that this resource no longer depends on another resource.
+   *
+   * This can be used for resources across stacks (or nested stack) boundaries
```
```suggestion
   * This can be used for resources across stacks (including nested stacks)
```

comcalvi commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -13,7 +13,7 @@ type Element = CfnResource | Stack;
  * - If there isn't a common stack, it means the elements belong to two
  *   disjoined stack-trees and therefore we apply the dependency at the
- *   assembly/app level between the two topl-level stacks.
+ *   assembly/app level between the two top-level stacks.
```
Good catch!

comcalvi commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -310,6 +335,24 @@ export class CfnResource extends CfnRefElement {
   }
 
+  /**
+   * Get a shallow copy of dependencies between this resource and other resources
+   * in the same stack.
+   */
+  public obtainResourceDependencies() {
+    return Array.from(this.dependsOn.values());
```
why not just return 

`this.dependsOn.values()` here? Is that not a shallow copy?

jusdino replied:
I'm pretty new to TypeScript (and jsii especially) but here's my understanding - please correct me if I'm wrong. Because `this.dependsOn` is a `Set`, we want to convert that into an `Array`, so that jsii can serialize it to other languages. To do that, we grab `.values()` which gets us an iterator, pass that to `Array.from()` which then constructs us a list. This has the happy side-effect of also giving us a 'shallow copy', so we aren't exposing a mutable container directly to a user.

comcalvi commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -261,6 +261,31 @@ export class CfnResource extends CfnRefElement {
+  /**
+   * Retrieves an array of resources this resource depends on.
+   *
+   * This assembles dependencies on resources across stacks (or nested stack)
+   * boundaries automatically.
```
```suggestion
   * This assembles dependencies on resources across stacks (including nested stacks)
   * automatically.
```

comcalvi commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -24,7 +24,48 @@ type Element = CfnResource | Stack;
+
+/**
+ * Find the appropriate location for a dependency and add or remove it
+ * @param operation
+ * @param source
+ * @param target
+ * @param reasonDescription
```
add descriptions to these, or remove these `@param` tags entirely.

comcalvi commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -70,31 +124,68 @@ export function addDependency<T extends Element>(source: T, target: T, reason?:
+
+/**
+ * Get a list of all resource-to-resource dependencies assembled from this Element, Stack or assembly-dependencies
+ * @param source
```
add a description or remove this tag entirely. 

If the descriptions become redundant or are too similar with descriptions of the same parameter in different method signatures, please delete the copies and just leave the one nearest to the top of the file, unless they're externally facing.

comcalvi commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -70,31 +124,68 @@ export function addDependency<T extends Element>(source: T, target: T, reason?:
+
+/**
+ * Find the resource in a common stack that 'points' to the given element
+ * @param element
+ * @param commonStack
```
same comment as above goes here.

comcalvi commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -70,31 +124,68 @@ export function addDependency<T extends Element>(source: T, target: T, reason?:
+function resourceInCommonStackFor(element: CfnResource | Stack, commonStack: Stack): CfnResource {
+  const resource = Stack.isStack(element) ? element.nestedStackResource : element;
+  if (!resource) {
+    throw new Error('assertion failure'); // see "assertion" above
```
this is not a very descriptive error message.

comcalvi commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -739,32 +741,114 @@ export class Stack extends CoreConstruct implements ITaggable {
+    }
+
+    if (process.env.CDK_DEBUG_DEPS) {
+      // eslint-disable-next-line no-console
+      console.error(`[CDK_DEBUG_DEPS] stack "${this.node.path}" no longer depends on "${target.node.path}" because: ${reasonFilter}`);
```
should be `throw new Error()`

jusdino replied:
I'm not sure about this one - this is intended to be debug message for troubleshooting. Changing this to throw an exception would break this method whenever the debug flag is set. Maybe it would be more appropriate to change this to `console.log()`, since this is not an error?

comcalvi replied:
yes, this should be a log call, not an error.

comcalvi reviewed (changes requested):
I like the additions this makes. However, if the intended use case is for users to write 

```
foo = stack.tryFindChild('Foo');
if (foo.obtainDependsOn().contains(bar)) {
  foo.removeDependsOn(bar);
fi
foo.addDependsOn(barReplacement);
```

we should also add a `replaceDependsOn(bar, barReplacement)` that does exactly what the above snippet does.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3c3f5561d6e3e64a7200da727e7f6b7834a86a76
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b23d54cf-2d37-4c56-84ce-c90560af3615%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ccfd779c3f2dc88e4fc8b9e6c7e0c82bef264f63
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=aa400887-a233-4018-8b73-c15a5617edac%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jusdino commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -190,6 +185,8 @@ function resourceInCommonStackFor(element: CfnResource | Stack, commonStack: Sta
 /**
  * Return a string representation of the given assembler, for use in error messages
+ *
+ * @internal
```
I have no idea how that got there :flushed:

jusdino commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -739,32 +741,114 @@ export class Stack extends CoreConstruct implements ITaggable {
+      reasonFilter.target = target;
+    }
+
+    let dep = this._stackDependencies[Names.uniqueId(target)];
```
I'm hoping that `options={}` is sensible for this case.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: e066b02b9044fed65f710b8281fa6a84088b918a
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=60c5ec4d-f57d-4600-8980-e0e30a178d57%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: c701ed0605ce2fe146f8eebcadb123fdd569679d
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bde1e5a4-3112-4f03-899f-29524f012958%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra reviewed (changes requested):
These test failures have to do with a bug we had when you last merged from main. Please rebase so we can see if the build is succeeding before we can provide another round of reviews.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 561f0a6acfac96be1f105df70003839e4afc5db4
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1467bcd2-b753-428a-b428-5427f53cd567%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
@jusdino Go into the aws-cdk-lib package and run `yarn pkglint`. Commit the change to the README file for aws-cdk-lib.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9aeaec9f3c672e2b9a4035d9c297ea462ff4078b
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3c82f85d-7c62-4be2-aea8-8eae43330c59%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -734,32 +736,113 @@ export class Stack extends Construct implements ITaggable {
+      reasonFilter.target = target;
+    }
+
+    let dep = this._stackDependencies[Names.uniqueResourceName(target, {})];
```
While @comcalvi is right that we're moving in this direction, I'm not sure it will work in this context. Because we're still using `uniqueId` throughout the rest of this, if you're looking for their unique resource name instead here, it may come up empty. I I think this exemplifies the need for integration tests on this change.

TheRealAmazonKendra replied:
Oh, also, you can totally just do `Names.uniqueResourceName(target)` if/when using this and omit the `{}` altogether. It's optional.

jusdino replied:
Ok, I'm working on some better tests, including integration as well as thinking through the `uniqueResourceName` vs `uniqueId` issue.

It looks like I'll have to follow the `@aws-cdk/core/test/integration.custom-resources.readme` pattern and actually place it under `@aws-cdk/aws-cloudformation`.

Also - TypeScript doesn't seem to agree about `options` being optional:
```lib/stack.ts:819:45 - error TS2554: Expected 2 arguments, but got 1.

819     let dep = this._stackDependencies[Names.uniqueResourceName(target)];
                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~

  lib/names.ts:84:59
    84   public static uniqueResourceName(construct: IConstruct, options: UniqueResourceNameOptions) {
                                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    An argument for 'options' was not provided.
```

TheRealAmazonKendra replied:
Lol, I wrote that code and don't properly remember what I actually did there. You'e right on this.

TheRealAmazonKendra reviewed (changes requested):
@comcalvi is already providing great feedback but I just want to add that I think this needs more testing. I'd like to see both more unit test coverage for these changes and integration tests as well. If you have any questions about writing integration tests, see our [integration test guide](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md).

jusdino commented:
Ok, I detected the `Names.uniqueId` bug in an existing unit test and added a two-part integration test. I can look into adding some more unit tests for further coverage next.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a24c5777f9fcfd6c953618568425924d822805a8
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d0a05874-8f8e-437e-b120-6a98e1434033%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 65bc9a5056c2904bd00ff6dc5655202a6c8c8f48
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8d49e3cf-8d46-41ba-98fb-fa8401fbfad1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 80516d2a7a1ba4cbba3a038a0df357e79356235a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=df2f6e0d-c73a-45d6-9fec-ecfe0473cff9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 07912f2dd68d97cc023c682dd404421fe2b0d0c5
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d9c93512-842a-4162-80b0-5e1b24504804%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: e71c4058014c18f6f15a754c8d6f435e365bca49
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bf8643a3-0773-4d62-bf42-513cc4e091dc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented on packages/@aws-cdk/aws-cloudformation/test/integ.core-deps.ts:
```diff
@@ -0,0 +1,78 @@
+import * as integ from '@aws-cdk/integ-tests';
+import { Construct } from 'constructs';
+
+/* eslint-disable @aws-cdk/no-core-construct */
```
What causes this violation?

jusdino replied:
copy-pasta :eyes:

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -15,6 +15,10 @@ import { TagManager } from './tag-manager';
 import { FeatureFlags } from './feature-flags';
+import { Stack } from './stack';
+
+type Element = CfnResource | Stack;
```
I'm not sure this will play nice with jsii.

jusdino replied:
I copied that from `deps.ts` since it seemed convenient, so it looks like it may work OK. If you confirm you want it gone, I'm happy to remove.

jusdino replied:
I'm still working on removing the `Element` types - I'm struggling a bit with not letting the type bleed out of `deps.ts` into the other interfaces.

jusdino replied:
I found this a useful reference to explain the jsii issue, btw:
https://aws.github.io/jsii/specification/2-type-system/#type-unions

> Type Unions
>
>In certain cases, several different kinds of values are acceptable for a given parameter or return type. TypeScript models those cases using type unions, which are represented as TypeA | TypeB. The jsii type model supports those, however most other statically typed languages do not have such a concept, making those parameters or return values difficult to use from those languages, as the value has to be declared using the most generic reference type available (for example, in Java, those are returned as java.lang.Object).
>
>When used as inputs (parameters, or properties of a struct), it may be possible to generate method overloads that will allow for a convenient API in languages that support overloads.
>
>In general however, type unions are discouraged and should only be used when there is no alternative way to model the API.

jusdino replied:
Okay, I had to be a bit more aggressive about replacing `Element` than I initially hoped - I had to remove it entirely, including from `deps.ts` and replace it with a token interface to represent the same, but without the union type. I'm not entirely sure how this interface will wind up represented in docs and target languages so I may need help putting some polish this part - if you like the approach anyway.

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -36,20 +73,33 @@ export function addDependency<T extends Element>(source: T, target: T, reason?:
     const topLevelSource = sourcePath[0]; // first path element is the top-level stack
     const topLevelTarget = targetPath[0];
-    topLevelSource._addAssemblyDependency(topLevelTarget, reason);
+    let reason = { description: reasonDescription, source: source, target: target };
```
Given that you aren't changing these values, why not make this a const?

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -20,6 +20,8 @@ import { LogicalIDs } from './private/logical-id';
 import { resolve } from './private/resolve';
 import { makeUniqueId } from './private/uniqueid';
 
+type Element = CfnResource | Stack;
```
Here, too, I'm concerned about jsii handling this.

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -734,32 +736,113 @@ export class Stack extends Construct implements ITaggable {
    * @internal
    */
-  public _addAssemblyDependency(target: Stack, reason?: string) {
+  public _addAssemblyDependency(target: Stack, reason: StackDependencyReason = { description: '' }) {
```
Why is description blank?

jusdino replied:
Since `reason` used to be optional and `reason.description` is the closest thing to a replacement, I moved to providing a blank default value. I could default it to something like `${source.node.path} depends on ${target.node.path}`, though that doesn't really add any information, since the `source` and `target` are already included in `reason`, now.

TheRealAmazonKendra replied:
I think this needs to be modeled in a way that a blank string isn't the default input. Description should likely be optional where this is defined.

jusdino commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -734,32 +736,113 @@ export class Stack extends Construct implements ITaggable {
-        stack: target,
-        reasons: [],
-      };
+      dep = this._stackDependencies[Names.uniqueResourceName(target, {})] = { stack: target, reasons: [] };
```
I'll leave the `Names.uniqueId` vs `Names.uniqueResourceName` update in the unit test just to help keep a guard rail out there for the future migration.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4d674af65f8ae24c994cf7a652f7081d1988004f
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eebb5d36-7189-4860-99d2-8a53e6ee005b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented on packages/@aws-cdk/aws-cloudformation/test/integ.core-deps.ts:
```diff
@@ -0,0 +1,76 @@
+/// !cdk-integ pragma:ignore-assets
```
This is no longer a thing we use.

TheRealAmazonKendra commented on packages/@aws-cdk/aws-cloudformation/test/integ.core-deps.ts:
```diff
@@ -0,0 +1,76 @@
+    serviceRole.node.tryRemoveChild('DefaultPolicy');
+  }
+}
+
```
```suggestion
```

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -289,6 +293,45 @@ export class CfnResource extends CfnRefElement {
+   * This assembles dependencies on resources across stacks (including nested stacks)
+   * automatically.
+   */
+  public obtainDependsOn() : Element[] {
```
I'm not so sure about this naming. Can this be refined a bit more? I think I'd be unsure if it was obtaining the resources that depend on this one or if it's obtaining the resources that depend on it.

TheRealAmazonKendra replied:
Ohhh. Now I see that an existing addDependsOn functions is probably why these were named this way. Interestingly, I see people not really using that. I wonder if it's because of the odd phrasing of it. Maybe we deprecate that naming convention and use a new one? What do you think?

jusdino replied:
Yeah, I Like `*Dependency` much better as well. I will adjust.

jusdino replied:
So the only reason I can think of to avoid the `addDependency` name in particular is potential confusion with `Node.addDependency()`. Maybe matching the name makes sense, since the higher-level Construct has the higher level interface with a matching name then, but maybe that could also be confusing. What do you think?

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -289,6 +293,45 @@ export class CfnResource extends CfnRefElement {
+   * This can be used for resources across stacks (including nested stacks)
+   * and the dependency will automatically be removed from the relevant scope.
+   */
+  public removeDependsOn(target: CfnResource) : void {
```
I wrote these comments out of order but kind of the same feedback at below. I'm not sure I like the <action>DependsOn phrasing of each of these.

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -289,6 +293,45 @@ export class CfnResource extends CfnRefElement {
+      return;
+    }
+
+    removeDependency(this, target, `"${Node.of(this).path}" depends on "${Node.of(target).path}"`);
```
I think this is actually the much better naming convention.

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -70,31 +120,67 @@ export function addDependency<T extends Element>(source: T, target: T, reason?:
+ * Get a list of all resource-to-resource dependencies assembled from this Element, Stack or assembly-dependencies
+ * @param source The source resource/stack (the dependent)
+ */
+export function obtainDependencies(source: Element) {
```
I think the verb here and elsewhere should be `get` instead of `obtain`

jusdino replied:
I think jsii yelled at me for calling things `get*` so I started using `obtain` when I saw it used elsewhere.

jusdino replied:
Ah ok yeah, here's the error. This means we can't use `get*` method names at all, doesn't it?
```error JSII5000: Methods and properties cannot have names like "getXxx": those conflict with Java property getters. Rename "CfnResource.getDependencies"```

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -1344,9 +1427,26 @@ function generateExportName(stackExports: Construct, id: string) {
+}
+
+interface StackDependencyReason {
+  description : string;
```
```suggestion
  description?: string;
```

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -1344,9 +1427,26 @@ function generateExportName(stackExports: Construct, id: string) {
 interface StackDependency {
   stack: Stack;
-  reasons: string[];
+  reasons: StackDependencyReason[];
```
I'm not sure I get what benefit you're getting from this change.

jusdino replied:
So, in order add a more complete add/replace/remove dependency interface, we need a clearer way to automatically look up and potentially remove a given stack dependency reason. Consider a case where we have two stacks, 1, and 2, each with resources A, B and C, D respectively:
```yaml
Stack1:
  - ResourceA
  - ResourceB
Stack2
  - ResourceC
  - ResourceD
```

If user code does both `ResourceA.addDependsOn(ResourceC)` and `ResourceB.addDependsOn(RecourceD)`, then it will result in only one stack dependency: `Stack1` on `Stack2`, but with two independent instances of `StackDependencyReason`. We need an unambiguous way to track those reasons so that the new `removeDependsOn` can appropriately back those dependencies back out, if the user desires. So if user code later does `ResourceA.removeDependsOn(ResourceC)`, the method will resolve it to a stack-level dependency and, rather than remove the dependency, it will find the matching `source` and `target` in the `StackDependencyReason`, remove that reason, but leave the stack dependency in place, since it will still have the `source: ResourceB, target: ResourceD` dependency reason. If the user code also then does `ResourceB.removeDependsOn(ResourceD)`, the method will again resolve to the stack level, remove the corresponding `StackDependencyReason` then, seeing that there is no longer a reason for the dependency, remove the `Stack1` on `Stack2` dependency as well.

Does that make sense? Without a clear data structure for automated reason matching, appropriately tracking resource-to-resource dependencies across stacks, we really can't be intelligent about removal in the future.

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -13,18 +13,55 @@ type Element = CfnResource | Stack;
  * diagnostics
  */
-export function addDependency<T extends Element>(source: T, target: T, reason?: string) {
+export function addDependency<T extends Element>(source: T, target: T, reason: string='dependency added using stack.addDependency()') {
```
Looks like the changes in this file don't have corresponding unit tests. Please add those.

TheRealAmazonKendra commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -20,6 +20,8 @@ import { LogicalIDs } from './private/logical-id';
 import { resolve } from './private/resolve';
```
The changes in this stack are also missing unit tests. Please add these as well.

TheRealAmazonKendra commented on packages/@aws-cdk/core/test/integration.deps.readme:
```diff
@@ -0,0 +1,5 @@
++--------------------------------------------------------------------------+
+| Since cdk-integ depends on cdk which depends on @aws-cdk/core (as a "dev |
+| dependency"), this integration test has been added to the package        |
+| @aws-cdk/aws-cloudformation under `test/integ.core-deps.ts`              |
++--------------------------------------------------------------------------+
```
Nice, now when people point out to me in the future that they think they can't write integration tests because there's not a preexisting package in here, I can point at this.

jusdino commented:
https://github.com/aws/aws-cdk/pull/20419#discussion_r962101794

I'm still working on removing the `Element` types - I'm struggling a bit with not letting the type bleed out of `deps.ts` into the other interfaces.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: e73289939c6484fdd04d82f5f60e6eb6576bac5f
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a79148fc-12e6-4bb3-89a5-fb01329a3a79%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: c0487446436296685dd3770da6b5dffaf7b7cd83
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8c9c4310-934b-4c8e-a992-227640eafab9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jusdino commented:
Had to roll back an update merge since it seems to have introduced a build error - something about `cxapi` not having an expected attribute.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ff92526654251ffeca7dd972d0ed671b208b7611
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3681d98a-ab1e-437d-92aa-a75e9ad4745b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
<!--STALE PR-->
This PR has been in the BUILD FAILING state for 3 weeks, and looks abandoned. To keep this PR from being closed, please continue work on it. If not, it will automatically be closed in a week.

jusdino commented:
> This PR has been in the BUILD FAILING state for 3 weeks, and looks abandoned. To keep this PR from being closed, please continue work on it. If not, it will automatically be closed in a week.

@TheRealAmazonKendra any chance you can turn off this auto-close? Still waiting for feedback on this update before I go trolling through the whole codebase to update references for this proposed method name change.

Naumel commented:
@Mergifyio update

With regard to feedback: the build is failing. That needs to be fixed before all other steps.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9041ad33b4bf9cdc6ac8fb7ff528fab3d90593a7
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=790dd510-48ec-4b3a-8cb1-f4786013b2a7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

Naumel reviewed (changes requested):
Hello, for better visibility, I will restate here what I added in the comments:

Feedback has been provided already and the ask is to fix the failing build before further feedback is to be added.

jusdino commented:
> Hello, for better visibility, I will restate here what I added in the comments:
> 
> Feedback has been provided already and the ask is to fix the failing build before further feedback is to be added.

Understood - just buying myself some time to work on this

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b44e35fe303227e3e34006bae34251fd705f07da
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=657fb4ee-89b3-4156-8204-56db52090063%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b65561c5d5cdff9db93f9f0c7f13bfec7c4ac075
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0938e4dc-12d7-4572-8de0-6c09ebc9f92c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
<!--STALE PR-->
This PR has been in the MERGE CONFLICTS state for 3 weeks, and looks abandoned. To keep this PR from being closed, please continue work on it. If not, it will automatically be closed in a week.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 273a0bb5c1c7c202e1acb50034c044f065875413
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=07642905-0b96-4acd-817b-4b70f38388e0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 7efa8fa031de195d6c6c35413a3f6ed475e5e1a6
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9e105019-15df-4256-89b0-6c3dfb4976cd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jusdino commented:
@Naumel, @TheRealAmazonKendra , is this still on your radar? I hope somebody can get back to this before I have to to another round of fussing with merge conflicts.

comcalvi commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -13,18 +19,55 @@ type Element = CfnResource | Stack;
+ * @param reason Optional description to associate with the dependency for
+ * diagnostics
+ */
+export function removeDependency(source: IElement, target: IElement, reason: string='dependency added using stack.removeDependency()') {
```
why the default value for `reason`? It doesn't look like we need this here.

comcalvi replied:
The default value of `'dependency added using stack.removeDependency()'` is confusing me. I assume this is a typo, but this reinforces my point that I don't think this default value is necessary.

jusdino replied:
yeah, that's a mistake. Will fix.

comcalvi commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -772,32 +772,116 @@ export class Stack extends Construct implements ITaggable {
    *
    * @internal
    */
-  public _addAssemblyDependency(target: Stack, reason?: string) {
```
this is a breaking change to a public method (`reason: string` -> `reason: StackDependencyReason`)

comcalvi replied:
This is `@internal`, so maybe this is fine?

comcalvi commented on packages/@aws-cdk/aws-cloudformation/test/integ.core-deps.ts:
```diff
@@ -0,0 +1,74 @@
+/*
+ * Real replaceDependency use case to test
+ *
+ * TestStack verification steps:
+ * - Deploy with `--no-clean`
+ * - Verify that the CloudFormation stack LogRetention CfnResource dependencies list CustomPolicy, not DefaultPolicy
+ *
+ * TestNestedStack verification steps:
+ * - Deploy with `--no-clean`
+ * - Verify that Stack2 lists Stack1 in DependsOn
+ */
```
leftover?

jusdino replied:
Not strictly leftover - I was following [this test](https://github.com/aws/aws-cdk/pull/20419#issuecomment-1319277855) as an example, which spelled out exactly what they were looking at in the template to verify it was desireable.

comcalvi commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -16,6 +16,7 @@ import { Tokenization } from './token';
 import { capitalizePropertyNames, ignoreEmpty, PostResolveToken } from './util';
 import { FeatureFlags } from './feature-flags';
 
+
```
no need for the extra blank line

comcalvi commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -273,14 +274,24 @@ export class CfnResource extends CfnRefElement {
    * and the dependency will automatically be transferred to the relevant scope.
    */
-  public addDependsOn(target: CfnResource) {
+  public addDependency(target: CfnResource) {
```
Why are we  adding a new method instead of just changing the body of `addDependsOn`?

jusdino replied:
:confused: I originally did just change the body, but @TheRealAmazonKendra asked me to change the name in a previous review.

comcalvi commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -289,6 +300,45 @@ export class CfnResource extends CfnRefElement {
+      return;
+    }
+
+    removeDependency(this, target, `"${Node.of(this).path}" depends on "${Node.of(target).path}"`);
```
Why does this need a reason description at all? A source and a target should be enough.

comcalvi replied:
This provides an argument for the reason that is completely different from the default value of `removeDependency()`'s parameter, and I still don't see why either of them exist.

jusdino replied:
Yeah the reason description definitely doesn't need to exist at all and I'm happy to remove. I added that because, in the existing internal api, there was that `reason` string, which seemed to be diagnostic in nature, so I kept a string in the `DependencyReason` as an attempt to preserve that behavior.

comcalvi commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -2,8 +2,14 @@ import { CfnResource } from './cfn-resource';
+/**
+ * This interface is intended to represent an object that is either a Stack or a CfnResource
+ */
+export interface IElement extends IConstruct {
```
No need for empty interfaces...can we keep the original type `Element` here?

jusdino replied:
@TheRealAmazonKendra did not like the `Element` type, since it is a union type and doesn't play well with jsii. At her request, I went about removing it where I could.

comcalvi replied:
@TheRealAmazonKendra I think the union type `Element` is safe here, since it isn't exported; since we already had the `Element` type (and it provides better safety than just `IConstruct`, I'd prefer to keep it.

jusdino replied:
Happy to do any way you all decide but FYI:
> So the Element or IElement ends up being part of the public interface as the return type of CfnResource.obtainDependencies(), so that's why I moved away from the union type all the way back into deps.ts.
> 
> So far as the benefit of IElement over just accepting IConstruct, there isn't one as written. I added this to be a placeholder for a possibly more specific interface after feedback from AWS, since I think we may wish to polish that piece up a touch more. This interface is what I was referring to when I asked for help at the end of this conversation:
https://github.com/aws/aws-cdk/pull/20419#discussion_r944611432

[From down here](https://github.com/aws/aws-cdk/pull/20419#discussion_r1025747070)

comcalvi commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -36,20 +79,33 @@ export function addDependency<T extends Element>(source: T, target: T, reason?:
   if (sourceStage !== targetStage) {
     // eslint-disable-next-line max-len
-    throw new Error(`You cannot add a dependency from '${source.node.path}' (in ${describeStage(sourceStage)}) to '${target.node.path}' (in ${describeStage(targetStage)}): dependency cannot cross stage boundaries`);
+    throw new Error(`You cannot have a dependency from '${source.node.path}' (in ${describeStage(sourceStage)}) to '${target.node.path}' (in ${describeStage(targetStage)}): dependency cannot cross stage boundaries`);
```
can we leave the original error message?

jusdino replied:
We can, though when this method is used in the `remove` case, the original message may be a bit confusing.

comcalvi commented on packages/@aws-cdk/core/lib/stack.ts:
```diff
@@ -1373,12 +1457,28 @@ function generateExportName(stackExports: Construct, id: string) {
+function filterReasons(dependency: StackDependency, reasonFilter: StackDependencyReason) {
+  let reasons: Set<StackDependencyReason> = new Set();
+  dependency.reasons.forEach((reason) => {
+    if ((!reasonFilter.source || reasonFilter.source == reason.source) && (!reasonFilter.target || reasonFilter.target == reason.target)) {
+      reasons.add(reason);
```
why are we adding reasons with no `source` and no `target` to this list?

MrArnoldPalmer commented on packages/@aws-cdk/core/README.md:
```diff
@@ -429,6 +429,10 @@ A stack dependency has the following implications:
 
+### CfnResource Dependencies
+
+To make declaring dependencies between `CfnResource` objects easier, you can declare dependencies from one `CfnResource` object on another by using the `cfnResource1.addDependesOn(cfnResource2)` method. This method will work for resources both within the same stack and across stacks as it detects the relative location of the two resources and adds the dependency either to the resource or between  ...
```
```suggestion
To make declaring dependencies between `CfnResource` objects easier, you can declare dependencies from one `CfnResource` object on another by using the `cfnResource1.addDepedency(cfnResource2)` method. This method will work for resources both within the same stack and across stacks as it detects the relative location of the two resources and adds the dependency either to the resource or between the relevant stacks, as appropriate. If more complex logic is in needed, you can similarly remove, replace, or view dependencies between `CfnResource` objects with the `CfnResource` `removeDependsOn`, `replaceDependsOn`, and `obtainDependsOn` methods, respectively.
```

Just docs since this is deprecating `addDependsOn`

comcalvi reviewed (changes requested):
This PR makes a number of breaking changes and adds a ton of machinery to our dependency logic, which is already fairly complex, and I'm not sure that we need all of this. 

The main benefit of this new logic is the ability to remove dependencies. I don't understand why we can't just traverse the construct tree to remove the dependencies.

comcalvi commented:
Overall I don't understand why this can't be implemented by adding only

* `_removeAssemblyDependency()`
* `removeDependency()`

and have those methods mirror the implementation of their `add` counterparts + some helper methods to pull out the common code.

MrArnoldPalmer commented on packages/aws-cdk-lib/README.md:
```diff
@@ -460,6 +460,10 @@ A stack dependency has the following implications:
 
+### CfnResource Dependencies
+
+To make declaring dependencies between `CfnResource` objects easier, you can declare dependencies from one `CfnResource` object on another by using the `cfnResource1.addDependesOn(cfnResource2)` method. This method will work for resources both within the same stack and across stacks as it detects the relative location of the two resources and adds the dependency either to the resource or between  ...
```
```suggestion
To make declaring dependencies between `CfnResource` objects easier, you can declare dependencies from one `CfnResource` object on another by using the `cfnResource1.addDependency(cfnResource2)` method. This method will work for resources both within the same stack and across stacks as it detects the relative location of the two resources and adds the dependency either to the resource or between the relevant stacks, as appropriate. If more complex logic is in needed, you can similarly remove, replace, or view dependencies between `CfnResource` objects with the `CfnResource` `removeDependsOn`, `replaceDependsOn`, and `obtainDependsOn` methods, respectively.
```

Same as above

MrArnoldPalmer commented on packages/@aws-cdk/core/lib/deps.ts:
```diff
@@ -2,8 +2,14 @@ import { CfnResource } from './cfn-resource';
+/**
+ * This interface is intended to represent an object that is either a Stack or a CfnResource
+ */
+export interface IElement extends IConstruct {
```
What's the benefit of changing from the union here since this type is only referenced in internal interfaces? Also what's the downside of all of the methods just accepting `IConstruct`?

jusdino replied:
So the `Element` or `IElement` ends up being part of the public interface as the return type of `CfnResource.obtainDependencies()`, so that's why I moved away from the union type all the way back into `deps.ts`.

So far as the benefit of `IElement` over just accepting `IConstruct`, there isn't one as written. I added this to be a placeholder for a possibly more specific interface after feedback from AWS, since I think we may wish to polish that piece up a touch more. This interface is what I was referring to when I asked for help at the end of this conversation:
https://github.com/aws/aws-cdk/pull/20419#discussion_r944611432

MrArnoldPalmer reviewed (commented):
Just assisting in trying to push this along, overall this looks pretty good to me though I'll give others who have reviewed a chance to chime in again if they want to.

jusdino commented on packages/@aws-cdk/core/lib/cfn-resource.ts:
```diff
@@ -289,6 +300,45 @@ export class CfnResource extends CfnRefElement {
+   * This assembles dependencies on resources across stacks (including nested stacks)
+   * automatically.
+   */
+  public obtainDependencies() {
```
Strictly speaking, yes, we could make due with only add/remove methods, but I've found that, especially when developing construct libraries for my developers to leverage, the lack of visibility into certain items in the cdk api, like construct dependencies, makes for a lot of frustration when trying to develop something with more nuance to it, because this is effectively a write-only interface, where there is zero ability to inspect what is actually there, only to attempt to change it.
Without this `obtain` method, a user is forced to blindly apply a `removeDependency` to every node in an entire construct tree, if they aren't sure where a dependency might be that they need to remove. While that approach would probably serve most use cases in the strictest sense, I can imagine that sometimes more nuance could be desired.
With that in mind, I would ask the converse: What is the harm in allowing a user to actually view what they are touching? All told, the `obtain` logic added is really quite limited.

jusdino commented:
Thanks for the reviews, @comcalvi and @MrArnoldPalmer . I'll try to address the new comments not part of a PR review change requested here:

> This PR makes a number of breaking changes and adds a ton of machinery to our dependency logic, which is already fairly complex, and I'm not sure that we need all of this.

I don't see any breaking changes here - what are you referring to? The closest thing I see is the `addDependsOn` -> `addDependency` deprecation, which I did specifically as a change requested by Amazon. I'm happy to roll that back, if that is ultimately what you want.

> The main benefit of this new logic is the ability to remove dependencies. I don't understand why we can't just traverse the construct tree to remove the dependencies.

Traversing the construct tree to remove dependencies is exactly what this enables. With the existing code, there is no mechanism to do that. I'm not sure what your point is here. Can you help me understand?

> Overall I don't understand why this can't be implemented by adding only
>
>    _removeAssemblyDependency()
>    removeDependency()
>
> and have those methods mirror the implementation of their add counterparts + some helper methods to pull out the common code.

Ok, I'm trying to boil down where my approach differs from what you're suggesting, since I don't think it's far from what I did. Aside from adding `_removeAssemblyDependency()` and `removeDependency()` and adding the helper methods to pull out the common code, I also:
- Added an `obtainDependencies()` method [discusion about that piece here](https://github.com/aws/aws-cdk/pull/20419#discussion_r1024450811)
- Added a `replaceDependency()` method [discussion about that piece here](https://github.com/aws/aws-cdk/pull/20419#pullrequestreview-1008233131)
- Deprecated `addDependsOn()` in favor of `addDependency()` [disscussion about that piece here](https://github.com/aws/aws-cdk/pull/20419#discussion_r962102636)

> If able I would love more info on the use case for the feature as #20418 provides some theoretical stuff but have you run into this limitation using specific constructs in aws-cdk-lib.

Sure - as it happens, [this integration test](https://github.com/aws/aws-cdk/pull/20419/files#diff-fbe0bc5d655c78e3d5c7d5635199689c2a45306cd05ddd7f0558d6c2d03ba616R19) is actually based on a real use case I've dealt with. Let's say we have a security rule that prohibits the use of in-line IAM policies in favor of managed policies (the the security rule is what it is, so let's not get into whether that's actually a good idea). In order to alleviate the pain experienced by our developers using AWS-CDK, we implemented an `Aspect` pattern that will find and replace all in-line policies with a managed policy resource. The problem we ran into is that, if any existing resources (like `LogRetention`) have added a dependency on the inline policy that we removed, we have no graceful way to salvage the resulting Cfn template.

Did I miss anything? I'm happy to work the smaller changes each of you pointed out, though I hesitate to invest in the details when there seem to be more fundamental questions about this work that still need sorting. Thoughts?

MrArnoldPalmer commented:
> Sure - as it happens, the integration test TestStack at aws-cloudformation/test/integ.core-deps.ts is actually based on a real use case I've dealt with. Let's say we have a security rule that prohibits the use of in-line IAM policies in favor of managed policies (the the security rule is what it is, so let's not get into whether that's actually a good idea). In order to alleviate the pain experienced by our developers using AWS-CDK, we implemented an Aspect pattern that will find and replace all in-line policies with a managed policy resource. The problem we ran into is that, if any existing resources (like LogRetention) have added a dependency on the inline policy that we removed, we have no graceful way to salvage the resulting Cfn template.

Thank you that definitely helps!

comcalvi commented:
>I don't see any breaking changes here - what are you referring to? The closest thing I see is the addDependsOn -> addDependency deprecation, which I did specifically as a change requested by Amazon. I'm happy to roll that back, if that is ultimately what you want.

I missed that the method whose signature changed was internal, so disregard that.

>Traversing the construct tree to remove dependencies is exactly what this enables. With the existing code, there is no mechanism to do that. I'm not sure what your point is here. Can you help me understand?

Sorry I wasn't more clear; I do not understand this mechanism. The main thing I'm missing the usage of `filterReasons()`. What does this function do? Why do we need to filter any reasons at all? I see `filterReasons()` used in a few spots, and I don't see what this is doing. 

>Ok, I'm trying to boil down where my approach differs from what you're suggesting, since I don't think it's far from what I did. Aside from adding _removeAssemblyDependency() and removeDependency() and adding the helper methods to pull out the common code, I also:

The method additions and name changes are good, this comment was about the new mechanism. A comment block detailing this mechanism would be helpful.

jusdino commented:
> Sorry I wasn't more clear; I do not understand this mechanism. The main thing I'm missing the usage of filterReasons(). What does this function do? Why do we need to filter any reasons at all? I see filterReasons() used in a few spots, and I don't see what this is doing.

Ah ok. So the `filterReasons` function is the core of what I'm explaining here:
https://github.com/aws/aws-cdk/pull/20419#discussion_r963923393

Would adding a comment block with something along those lines be helpful?

comcalvi commented:
>the method will resolve it to a stack-level dependency and, rather than remove the dependency, it will find the matching source and target in the StackDependencyReason, remove that reason, but leave the stack dependency in place, since it will still have the source: ResourceB, target: ResourceD dependency reason. If the user code also then does ResourceB.removeDependsOn(ResourceD), the method will again resolve to the stack level, remove the corresponding StackDependencyReason then, seeing that there is no longer a reason for the dependency, remove the Stack1 on Stack2 dependency as well.

I don't see the purpose of filtering reasons in this process though. From this, I understand that we need to track resource dependencies across stacks. I think today we don't do this, and instead only track the stack itself as a dependency; this is a problem. As you explain, if we only track the stack as a dependency and a user removes one dependency but not the other, we can't determine that the stack dependency should actually be left in place; therefore, we need a mechanism to track the resources themselves as dependencies and not just the stack.

With that in mind, I would expect the fundamental change to add the construct path to the dependency list, eg `Stack2/Resource` (where today, it's just `Stack2`). To achieve that, you're tracking source and target for each dependency. That makes sense. What doesn't make sense to me is why we need to filter the dependency reasons. What problem does filtering dependencies sovle?

jusdino commented:
> What doesn't make sense to me is why we need to filter the dependency reasons. What problem does filtering dependencies sovle?

Sorry for the slow reply - I've been sick and wanted to think this through. My short answer is, looking closer, the `filterReasons` may be a bit overkill.

What we really need is to be able to match a `StackDependencyReason` reliably when removing resource-to-resource dependencies that ended up resolving as stack-to-stack dependencies. The `filterReasons` function is doing this for us but it may be a bit more complex than necessary specifically because I was trying to be a bit defensive in my approach. I was specifically thinking about the situation where, somehow, a `source` or `target` of `null` or `undefined` was added to the `stackDependency.reasons`. I don't think the public interface allows such a thing to happen, but if it did, I wanted a smarter approach to pulling out a match than specifically looking for `source == reason.source && target == reason.target`. Otherwise, that reason would never match and so would prevent the `StackDependency` from ever being completely removed.

It sounds like your preference would be to simplify. I could simplify/remove the `filterReasons` function if you would rather leave that null/undefined unaddressed, which I think is probably fine since it _shouldn't_ happen, anyway.

comcalvi commented:
Thanks for the explanation, this makes sense. I'd prefer we throw an error if we encounter a null` or `undefined` reason and remove the `filterReasons()` method entirely, since such a reason existing means that something has gone wrong.

jusdino commented:
> Thanks for the explanation, this makes sense. It's good to think of edge cases like this, and we should definitely guard against it. However I'd prefer to fail in this case, since the API shouldn't allow this to happen. Even it happens to be impossible for this case to occur with today's code, future changes may have bugs that create a `null` or `undefined` reason that could be silently ignored if we just filter out those reasons. I'd prefer we throw an error if we encounter a `null` or `undefined` reason and remove the `filterReasons()` method entirely, since such a reason existing means that something has gone wrong and the deployment may fail (or worse yet, deploy successfully by chance but with missing dependency logic which may cause it to fail for seemingly no reason in the future).

Sounds good. I'll work on that next along with the smaller changes from your last review.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 76f198632982500a78a2c910f7c89c7d67f49075
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0440b7a8-14e8-4e7c-a500-071ef782efef%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jusdino commented:
@comcalvi, @MrArnoldPalmer, I think this covers everything except for what to do about `IElement` - just need some direction on which way to go with that.

comcalvi commented:
for `IElement`, please leave the original union type `Element` as it was.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 18fa180123b117c07edd8fdfe065dbfba123263a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c8120ace-97c9-4497-8969-451af3b553be%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 508b9b683f76c43229b1babfb1e7ff3515e61cc5
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=16379c08-66a6-4473-92b7-33533be2af42%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

comcalvi commented:
@jusdino Can you please enable contributions by maintainers to your fork? I have a commit I'd like to push to this PR, and then I'll be happy to approve it. Thanks for all your hard work on this PR, this is in a really good spot now!

jusdino commented:
> @jusdino Can you please enable contributions by maintainers to your fork? I have a commit I'd like to push to this PR, and then I'll be happy to approve it. Thanks for all your hard work on this PR, this is in a really good spot now!

Glad to be close to the finish line!

So this PR is from a fork on a GitHub organization and I wasn’t able to find a way to do that with my last PR from there. Last time, I ended up reopening a new PR from a personal fork. I can do that here as well though it will take me a bit - got some personal stuff going on atm.

For reference:
https://github.com/aws/aws-cdk/pull/20419#issuecomment-1195662725

comcalvi commented:
Thanks! Just link it here and I'll approve once I push a commit to it.

jusdino commented:
@comcalvi , #23383 is ready :+1:

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 6ed2097cb3fb5557b88e6967a19ef47235fa6211
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2793f343-1f49-4e0b-82b0-00ad835bcf4c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

comcalvi commented:
Closing this one in favor of #23383
